### DISCUSSION TITLE
Use custom port makes FTP connection fail

### DISCUSSION THREAD

**Discussion by negstek** [apache/airflow#40239] [General]
### Apache Airflow version

2.9.2

### What happened?

When I use a custom port when I try to create a FTP connection, I got "530 Not logged in." when I test it.
Same problem after the creation if I try to use it in a DAG.

### What you think should happen instead?

the connexion should work fine.

### How to reproduce

create a FTP connection with a custom port in the web GUI. Try to test the connection in teh web GUI.

### Operating System

not relevant but linux in wsl

### Deployment

Docker-Compose

### Anything else?

maybe linked to this issue :
https://github.com/apache/airflow/pull/39465

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
I think this issue has not enough information to act on lt - some more details logs and precise explanation with screenshots what you've done would be needed for anyone who might want to take a look

**Comment by negstek** [comment]
> I think this issue has not enough information to act on lt - some more details logs and precise explanation with screenshots what you've done would be needed for anyone who might want to take a look

Sorry I thought it was pretty clear:
![image](https://github.com/apache/airflow/assets/883536/bd48d659-42b6-49ce-b3cf-90c895483b57)
```
airflow-webserver-1     | [2024-06-14 14:46:36 +0000] [14] [CRITICAL] WORKER TIMEOUT (pid:40)
airflow-webserver-1     | [2024-06-14 14:46:36 +0000] [40] [ERROR] Error handling request /api/v1/connections/test
airflow-webserver-1     | Traceback (most recent call last):
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/gunicorn/workers/sync.py", line 135, in handle
airflow-webserver-1     |     self.handle_request(listener, req, client, addr)
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/gunicorn/workers/sync.py", line 178, in handle_request
airflow-webserver-1     |     respiter = self.wsgi(environ, resp.start_response)
airflow-webserver-1     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/flask/app.py", line 2552, in __call__
airflow-webserver-1     |     return self.wsgi_app(environ, start_response)
airflow-webserver-1     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/flask/app.py", line 2529, in wsgi_app
airflow-webserver-1     |     response = self.full_dispatch_request()
airflow-webserver-1     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/flask/app.py", line 1823, in full_dispatch_request
airflow-webserver-1     |     rv = self.dispatch_request()
airflow-webserver-1     |          ^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/flask/app.py", line 1799, in dispatch_request
airflow-webserver-1     |     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
airflow-webserver-1     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/connexion/decorators/decorator.py", line 68, in wrapper
airflow-webserver-1     |     response = function(request)
airflow-webserver-1     |                ^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/connexion/decorators/uri_parsing.py", line 149, in wrapper
airflow-webserver-1     |     response = function(request)
airflow-webserver-1     |                ^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/connexion/decorators/validation.py", line 196, in wrapper
airflow-webserver-1     |     response = function(request)
airflow-webserver-1     |                ^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/connexion/decorators/response.py", line 112, in wrapper
airflow-webserver-1     |     response = function(request)
airflow-webserver-1     |                ^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/connexion/decorators/parameter.py", line 120, in wrapper
airflow-webserver-1     |     return function(**kwargs)
airflow-webserver-1     |            ^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/airflow/api_connexion/security.py", line 126, in decorated
airflow-webserver-1     |     return _requires_access(
airflow-webserver-1     |            ^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/airflow/api_connexion/security.py", line 98, in _requires_access
airflow-webserver-1     |     return func(*args, **kwargs)
airflow-webserver-1     |            ^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/airflow/api_connexion/endpoints/connection_endpoint.py", line 203, in test_connection
airflow-webserver-1     |     status, message = conn.test_connection()
airflow-webserver-1     |                       ^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/airflow/models/connection.py", line 463, in test_connection
airflow-webserver-1     |     status, message = hook.test_connection()
airflow-webserver-1     |                       ^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/airflow/providers/ftp/hooks/ftp.py", line 261, in test_connection
airflow-webserver-1     |     conn = self.get_conn()
airflow-webserver-1     |            ^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/airflow/providers/ftp/hooks/ftp.py", line 61, in get_conn
airflow-webserver-1     |     self.conn = ftplib.FTP(params.host, params.login, params.password)  # nosec: B321
airflow-webserver-1     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/usr/local/lib/python3.12/ftplib.py", line 121, in __init__
airflow-webserver-1     |     self.connect(host)
airflow-webserver-1     |   File "/usr/local/lib/python3.12/ftplib.py", line 158, in connect
airflow-webserver-1     |     self.sock = socket.create_connection((self.host, self.port), self.timeout,
airflow-webserver-1     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
airflow-webserver-1     |   File "/usr/local/lib/python3.12/socket.py", line 838, in create_connection
airflow-webserver-1     |     sock.connect(sa)
airflow-webserver-1     |   File "/home/airflow/.local/lib/python3.12/site-packages/gunicorn/workers/base.py", line 203, in handle_abort
airflow-webserver-1     |     sys.exit(1)
airflow-webserver-1     | SystemExit: 1
```

**Comment by potiuk** [comment]
Seems like you have a timeout which indicate either server side issue or firewall. So not airflow issue. converting into a discussion.

**Comment by negstek** [comment]
@potiuk I think you are wrong I got near a hundred of working connection in FTP and SFTP and only 3 FTP fails, that were the connection with custom port, I made tests in filezilla, all is working in my network. More of that the implementation of  ftplib in airflow not work. The port is not set in the stacktrace:

```
`File "/home/airflow/.local/lib/python3.12/site-packages/airflow/providers/ftp/hooks/ftp.py", line 61, in get_conn
airflow-webserver-1     |     self.conn = ftplib.FTP(params.host, params.login, params.password)
 ```
 
 this kind of code does not set the port you should do it like this in ftplib:
 
 ```
 from ftplib import FTP
 
 host = ""
login = ""
password = ""

conn = FTP()
conn.connect(host, port)
conn.login(login, password)
```

If you read carefully my ticket you have seen that same kind of bug have been fixed in FTPHook 3 weeks ago at the "Anything else?" section

maybe linked to this issue :
https://github.com/apache/airflow/pull/39465

**Reply by potiuk** [reply]
Can you test it with the latest version? If it's been fixed, then there is even less need to create a duplicate issue for it.

**Reply by potiuk** [reply]
(also remember that you can upgrade providers independently from Airflow - so you can test latest version of a provider easily -the docs at installation page explain how to upgrade providers independently from Airflow.

**Reply by potiuk** [reply]
BTW. I think you also might be wrong:

`self.sock = socket.create_connection((self.host, self.port), self.timeout,` actually sets the port.

**Comment by negstek** [comment]
@potiuk I am using version 2.9.2 as you can see in the header of my ticket, I just download it today with the hope that FTPHook fix also fixmy bug, is there another one ?

thank you for support BTW I will try to take a look at dependencies monday, have a good week.

**Reply by potiuk** [reply]
All information can be found in release notes.

As I mentioned - providers are independent from Airflow. see the docs. You can check release notes in FTP provider if the version you have has a fix, if not you wil have to wait for it to be released (see README for our release approach for providers) , or you should build it yourself following CONTRIBUTION guide if you are impatient and want to help by testing it from main.

Generally speaking we always appreciate if our users contribute back for the free software they get, by going above and beyond and testing stuff.

If the fix is not yet released you can subscribe to the devlist (you will find link in the community section of our docs) and when the new version of provider's RC is released we ask our users to help with testing - so you will be able to test RC then.

Those are generally all options and direction to the docs that might help you in assessing where things are.
